netutils/dhcp6c/dhcp6c.c

```diff
@@ -0,0 +1,1951 @@
+    }
+
+  return 0;
+}
+
+static int dhcp6c_commit_advert(FAR void *handle, uint32_t elapsed)
+{
+  FAR struct dhcp6c_state_s *pdhcp6c = (FAR struct dhcp6c_state_s *)handle;
+  size_t cand_len;
+  FAR struct dhcpv6_server_cand_s *c = NULL;
+  FAR struct dhcpv6_server_cand_s *cand = dhcp6c_get_state(handle,
+                                                           STATE_SERVER_CAND,
+                                                           &cand_len);
+  bool retry = false;
+
+  for (size_t i = 0; i < cand_len / sizeof(*c); ++i)
+    {
+      if (cand[i].has_noaddravail)
+        {
+          retry = true;
+        }
+
+      if (c == NULL || c->preference < cand[i].preference)
+        {
+          c = &cand[i];
+        }
+    }
+
+  if (retry && pdhcp6c->ia_mode == IA_MODE_TRY)
+    {
+      /* We give it a second try without the IA_NA */
+
+      pdhcp6c->ia_mode = IA_MODE_NONE;
+      return dhcp6c_command(handle, DHCPV6_MSG_SOLICIT);
+    }
+
+  if (c != NULL)
+    {
+      uint16_t hdr[2] =
+      {
+        htons(DHCPV6_OPT_SERVERID), htons(c->duid_len)
+      };
+
+      dhcp6c_add_state(handle, STATE_SERVER_ID, hdr, sizeof(hdr));
+      dhcp6c_add_state(handle, STATE_SERVER_ID, c->duid, c->duid_len);
+      pdhcp6c->accept_reconfig = c->wants_reconfigure;
+    }
+
+  dhcp6c_clear_state(handle, STATE_SERVER_CAND);
+
+  if (c == NULL)
+    {
+      return -1;
+    }
+  else if (pdhcp6c->request_prefix || pdhcp6c->ia_mode != IA_MODE_NONE)
+    {
+      return DHCPV6_STATEFUL;
+    }
+  else
+    {
+      return DHCPV6_STATELESS;
+    }
+}
+
+static time_t dhcp6c_parse_ia(FAR void *handle, FAR void *opt, FAR void *end)
+{
+  FAR struct dhcp6c_state_s *pdhcp6c = (FAR struct dhcp6c_state_s *)handle;
+  uint32_t timeout = UINT32_MAX;
+  uint16_t otype;
+  uint16_t olen;
+  uint16_t stype;
+  uint16_t slen;
+  FAR uint8_t *ite0;
+  FAR uint8_t *odata;
+  FAR uint8_t *sdata;
+
+  /* Update address IA */
+
+  dhcpv6_for_each_option(ite0, opt, end, otype, olen, odata)
+    {
+      if (otype == DHCPV6_OPT_IA_PREFIX)
+        {
+          FAR struct dhcpv6_ia_prefix_s *prefix = (FAR void *)(odata - 4);
+          FAR struct dhcpv6_ia_prefix_s *local = NULL;
+          uint32_t valid;
+          uint32_t pref;
+          size_t pd_len;
+          FAR uint8_t *pd;
+          FAR uint8_t *ite1;
+
+          if (olen + 4u < sizeof(*prefix))
+            {
+              continue;
+            }
+
+          olen = sizeof(*prefix);
+          valid = ntohl(prefix->valid);
+          pref = ntohl(prefix->preferred);
+
+          if (pref > valid)
+            {
+              continue;
+            }
+
+          /* Search matching IA */
+
+          pd = dhcp6c_get_state(handle, STATE_IA_PD, &pd_len);
+          dhcpv6_for_each_option(ite1, pd, pd + pd_len,
+                                 stype, slen, sdata)
+            {
+              if (memcmp(sdata + 8, odata + 8,
+                         sizeof(local->addr) + 1) == 0)
```

Comment:
Why we have `+1` here?

Reply:
prefix address including length and value.
length occupies one byte